core/src/main/scala/org/apache/spark/util/CheckpointingIterator.scala

```diff
@@ -0,0 +1,225 @@
+    if (serializeStream != null) {
+      serializeStream.close()
+    }
+    fs.delete(tempOutputPath, false)
+  }
+
+  override def hasNext: Boolean = {
+    try {
+      val r = values.hasNext
+      if (!r) {
+        complete()
+      }
+      r
+    } catch {
+      case e: Throwable =>
+        try {
+          cleanup()
+        } catch {
+          case NonFatal(e1) =>
+            // Log `e1` since we should not override `e`
+            logError(e1.getMessage, e1)
+        }
+        throw e
+    }
+  }
+
+  override def next(): T = {
+    try {
+      val value = values.next()
+      serializeStream.writeObject(value)
+      value
+    } catch {
+      case e: Throwable =>
+        try {
+          cleanup()
+        } catch {
+          case NonFatal(e1) =>
+            // Log `e1` since we should not override `e`
+            logError(e1.getMessage, e1)
+        }
+        throw e
+    }
+  }
+}
+
+private[spark] object CheckpointingIterator {
+
+  private val checkpointingRDDPartitions = new ConcurrentHashMap[RDDBlockId, RDDBlockId]()
+
+  /**
+   * Return true if the caller gets the lock to write the checkpoint file. Otherwise, the caller
+   * should not do checkpointing.
+   */
+  private def acquireLockForPartition(id: RDDBlockId): Boolean = {
+    checkpointingRDDPartitions.putIfAbsent(id, id) == null
+  }
+
+  /**
+   * Release the lock to avoid memory leak.
+   */
+  private def releaseLockForPartition(id: RDDBlockId): Unit = {
+    checkpointingRDDPartitions.remove(id)
+  }
+
+  def apply[T: ClassTag](
+      rdd: RDD[T],
+      values: Iterator[T],
+      path: String,
+      broadcastedConf: Broadcast[SerializableConfiguration],
+      partitionIndex: Int,
+      context: TaskContext,
+      blockSize: Int = -1): Iterator[T] = {
+    val id = RDDBlockId(rdd.id, partitionIndex)
+    if (CheckpointingIterator.acquireLockForPartition(id)) {
+      try {
+        val outputDir = new Path(path)
+        val fs = outputDir.getFileSystem(broadcastedConf.value.value)
+        val finalOutputName = ReliableCheckpointRDD.checkpointFileName(partitionIndex)
+        val finalOutputPath = new Path(outputDir, finalOutputName)
+        if (fs.exists(finalOutputPath)) {
+          // RDD has already been checkpointed by a previous task. So we don't need to checkpoint
+          // again.
+          return values
+        }
+        val tempOutputPath =
+          new Path(outputDir, "." + finalOutputName + "-attempt-" + context.attemptNumber)
+        if (fs.exists(tempOutputPath)) {
+          throw new IOException(s"Checkpoint failed: temporary path $tempOutputPath already exists")
+        }
+
+        new CheckpointingIterator(
+          id,
+          values,
+          fs,
+          tempOutputPath,
+          finalOutputPath,
+          context,
+          blockSize)
+      } catch {
+        case e: Throwable =>
+          releaseLockForPartition(id)
+          throw e
+      }
+    } else {
+      values
```

Comment:
add a comment here `// Iterator is already checkpointed, so just return the values`